src/components/lib/cos_kernel_api.c

```diff
@@ -300,26 +303,78 @@ __page_bump_valloc(struct cos_compinfo *ci)
 		if (call_cap_op(meta->captbl_cap, CAPTBL_OP_PGTBLACTIVATE,
 				pte_cap, meta->mi.pgtbl_cap, ptemem_cap, 1)) {
 			assert(0); /* race? */
-			return 0;
 		}
 
 		/* Construct pgtbl */
-		if (call_cap_op(ci->pgtbl_cap, CAPTBL_OP_CONS, pte_cap, heap_vaddr, 0, 0)) {
+		if (call_cap_op(cipgtbl, CAPTBL_OP_CONS, pte_cap, addr, 0, 0)) {
 			assert(0); /* race? */
-			return 0;
 		}
+	}
 
-		ci->vasrange_frontier += PGD_RANGE;
-		assert(ci->vasrange_frontier == round_up_to_pgd_page(ci->vasrange_frontier));
+	/* assert(round_up_to_pgd_page(addr) == round_up_to_pgd_page(mem_ptr + mem_sz)); */
+	
+	return mem_ptr;
+}
+
+static void
+__cos_meminfo_alloc(struct cos_compinfo *ci, vaddr_t untyped_ptr, unsigned long untyped_sz)
+{
+	vaddr_t addr;
+	struct cos_compinfo *meta = __compinfo_metacap(ci);
+
+	/*
+	 * FIXME: untyped_ptr and untyped_sz better be PGD_RANGE aligned.
+	 */
+	assert(__bump_mem_expand_range(ci, ci->mi.pgtbl_cap, untyped_ptr, untyped_sz) == untyped_ptr);
+	
+	vaddr_t start_addr        = meta->mi.untyped_frontier - untyped_sz;
+	meta->mi.untyped_frontier = start_addr;
+
+	for (addr = untyped_ptr; addr < untyped_ptr + untyped_sz; addr += PAGE_SIZE, start_addr += PAGE_SIZE) {
+		if (call_cap_op(meta->mi.pgtbl_cap, CAPTBL_OP_MEMMOVE, start_addr, ci->mi.pgtbl_cap, addr, 0)) BUG();
+	}
+}
+
+void
+cos_meminfo_alloc(struct cos_compinfo *ci, vaddr_t untyped_ptr, unsigned long untyped_sz)
```

Comment:
It mainly allocates untyped and updates the meminfo with the arguments passed, again wasn't sure what would be the right name, untyped_alloc or meminfo_alloc. Because functionally it does what meminfo_init does and other than allocating untyped memory, so I kept it this way.